plumbing/format/diff/unified_encoder.go

```diff
@@ -0,0 +1,345 @@
+type UnifiedEncoder struct {
+	io.Writer
+	CtxLines int
+}
+
+func (e *UnifiedEncoder) Encode(patch Patch) error {
+	if err := e.printMessage(patch.Message()); err != nil {
+		return err
+	}
+
+	for _, p := range patch.FilePatches() {
+		f, t := p.Files()
+		if err := e.header(f, t, p.IsBinary()); err != nil {
+			return err
+		}
+
+		g := newHunksGenerator(p.Chunks(), e.CtxLines)
+		for _, c := range g.Generate() {
+			if err := e.prints(c.String()); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (e *UnifiedEncoder) printMessage(message string) error {
+	isEmpty := message == ""
+	hasSuffix := strings.HasSuffix(message, "\n")
+	msg := message
+	if !isEmpty && !hasSuffix {
+		msg = msg + "\n"
+	}
+
+	return e.prints(msg)
+}
+
+func (e *UnifiedEncoder) header(from, to File, isBinary bool) error {
+	var fromPath, toPath string
+	var initLine, modesLine, indexLine string
+	var equals bool
+
+	switch {
+	case from == nil && to == nil:
+		return ErrBothFilesEmpty
+	case from != nil && to != nil:
+		initLine = fmt.Sprintf(diffInit, from.Path(), to.Path())
+		if from.Mode() != to.Mode() {
+			modesLine = fmt.Sprintf(oldMode+newMode, from.Mode(), to.Mode())
+			indexLine = fmt.Sprintf(indexNoMode, from.Hash(), to.Hash())
+		} else {
+			indexLine = fmt.Sprintf(indexAndMode, from.Hash(), to.Hash(), from.Mode())
+		}
+
+		if from.Path() != to.Path() {
+			modesLine += fmt.Sprintf(
+				renameFileMode+renameFileMode,
+				renameFrom, from.Path(), renameTo, to.Path())
+		}
+
+		if from.Hash() == to.Hash() {
+			equals = true
+		}
+
+		fromPath = aDir + from.Path()
+		toPath = bDir + to.Path()
+	case from == nil:
+		initLine = fmt.Sprintf(diffInit, to.Path(), to.Path())
+		modesLine = fmt.Sprintf(newFileMode, to.Mode())
+		indexLine = fmt.Sprintf(indexNoMode, plumbing.ZeroHash, to.Hash())
+
+		fromPath = noFilePath
+		toPath = bDir + to.Path()
+	case to == nil:
+		initLine = fmt.Sprintf(diffInit, from.Path(), from.Path())
+		modesLine = fmt.Sprintf(deletedFileMode, from.Mode())
+		indexLine = fmt.Sprintf(indexNoMode, from.Hash(), plumbing.ZeroHash)
+
+		fromPath = aDir + from.Path()
+		toPath = noFilePath
+	}
+
+	if equals {
+		return e.prints(initLine + modesLine)
+	}
+
+	format := fPath + tPath
+	if isBinary {
+		format = binary
+	}
+	pathLines := fmt.Sprintf(format, fromPath, toPath)
+
+	return e.prints(initLine + modesLine + indexLine + pathLines)
+}
+
+func (e *UnifiedEncoder) prints(s string) error {
+	_, err := e.Write([]byte(s))
+	return err
+}
+
+type hunksGenerator struct {
+	fromLine, toLine            int
+	ctxLines                    int
+	chunks                      []Chunk
+	current                     *hunk
+	hunks                       []*hunk
+	beforeContext, afterContext []string
+}
+
+func newHunksGenerator(chunks []Chunk, ctxLines int) *hunksGenerator {
+	return &hunksGenerator{
+		chunks:   chunks,
+		ctxLines: ctxLines,
+	}
+}
+
+func (c *hunksGenerator) Generate() []*hunk {
```

Comment:
Because the struct is private, to know which methods are intended to be used outside.